### PR TITLE
Enable to use input_size at transforms in recipe

### PR DESCRIPTION
**Pull request by eunwoosh**
### Summary
This PR enables to use input_size at transforms in recipe.
It's for unified input size. So, if value of all init_args in recipe uses input_size, all necessary values are automatically updated by changing only input_size.
For example, let's assume that we have recipe as below.
```
  data:
    train_subset:
      transforms:
        - class_path: otx.core.data.transform_libs.torchvision.RandomResize
          init_args:
            scale
            - 1000
            - 1000
        - class_path: otx.core.data.transform_libs.torchvision.RandomCrop
          init_args:
            crop_size
            - 500
            - 500
```
It can be changed as below using input_size.
```
  data:
    input_size:
    - 1000
    - 1000
    train_subset:
      transforms:
        - class_path: otx.core.data.transform_libs.torchvision.RandomResize
          init_args:
            scale: $(inputs_size)
        - class_path: otx.core.data.transform_libs.torchvision.RandomCrop
          init_args:
            crop_size: $(input_size) * 0.5
```
As you can see above, `$(input_size)` should be used to notify to use input_size.
Only simple multiplication or division evaluation is supported. So, addition or subtraction can't be used.
Type of input_size can be `int` or `tuple[int, int]`. If `int` value is given, it's considered as `tuple[int, int]` with same value.
whether passing `int` or `tuple[int, int]` to transform class is decided based on type hint. If there is no type hint, `tuple[int, int]` is passed.

input_size can be overrided by subset config. For example, if we want to use (1000, 1000) input size at train and (800, 800) at validation, recipe below can be used.
```
  data:
    input_size:
    - 1000
    - 1000
    train_subset:
      transforms:
        - class_path: otx.core.data.transform_libs.torchvision.RandomResize
          init_args:
            scale: $(inputs_size)
    val_subset:
      input_size:
      - 800
      - 800
      transforms:
        - class_path: otx.core.data.transform_libs.torchvision.RandomResize
          init_args:
            scale: $(inputs_size)
```
Of course, setting input_size only in subset confg rather than data config is available.
But input_size can't use `$(input_size)`. It's available only in `transforms`.

### Checklist

- [x] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have ran e2e tests and there is no issues.
- [x] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2024 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

File: src/otx/core/utils/utils.py [hunk 1]
```diff
@@ -84,3 +85,10 @@ def get_idx_list_per_classes(dm_dataset: DmDataset, use_string_label: bool = Fal
     for k in stats:
         stats[k] = list(dict.fromkeys(stats[k]))
     return stats
+
+
+def get_obj_from_str(obj_path: str) -> Any:  # noqa: ANN401
+    """Get object from import format string."""
+    module_name, obj_name = obj_path.rsplit(".", 1)
+    module = importlib.import_module(module_name)
+    return getattr(module, obj_name)
```
Comment (goodsong81) [inline src/otx/core/utils/utils.py]: Function name seems not telling what it does; `import_object_from_module()`?

File: src/otx/core/data/transform_libs/torchvision.py [hunk 2]
```diff
@@ -2364,7 +2369,7 @@ class RandomCrop(tvt_v2.Transform, NumpytoTVTensorMixin):
 
     def __init__(
         self,
-        crop_size: tuple,  # (H, W)
+        crop_size: tuple[int, int],  # (H, W)
```
Comment (goodsong81) [inline src/otx/core/data/transform_libs/torchvision.py:2372]: Is this always true in OTX? No (W, H)? OK if it is so but we'd better modify test cases for non-square input size to check for the future changes.

File: CHANGELOG.md [hunk 3]
```diff
@@ -10,6 +10,8 @@ All notable changes to this project will be documented in this file.
 
 - Enable to override data configurations
   (<https://github.com/openvinotoolkit/training_extensions/pull/3748>)
+- Enable to use input_size at transforms in reipce
```
Comment (sungchul1) [inline CHANGELOG.md]:
```suggestion
- Enable to use input_size at transforms in recipe
```

File: tests/unit/core/data/test_module.py [hunk 4]
```diff
@@ -115,13 +119,17 @@ def test_init(
             train_subset=fxt_config.train_subset,
             val_subset=fxt_config.val_subset,
             test_subset=fxt_config.test_subset,
+            input_size=input_size,
```
Comment (harimkang) [inline tests/unit/core/data/test_module.py]: Maybe we need an additional test to see if the overriding works depending on the position of `input_size`?
Also, could there be a test case to see if it works if `input_size` is None and each transform has a corresponding value?

It would be nice to have a test case to verify that it works in the two cases below.
1. `input_size` is None & `each_subset.input_size` is different.
2. input_size` is None & `each_subset.input_size` is None & `each_transform.scale` is not None.

File: src/otx/core/data/transform_libs/torchvision.py [hunk 5]
```diff
@@ -3119,11 +3124,100 @@ def generate(cls, config: SubsetConfig) -> Compose:
 
         transforms = []
         for cfg_transform in config.transforms:
+            if (input_size := getattr(config, "input_size", None)) is not None:
```
Comment (sungchul1) [inline src/otx/core/data/transform_libs/torchvision.py]: Do we need to get `input_size` every iteration?

File: src/otx/core/data/transform_libs/torchvision.py [hunk 6]
```diff
@@ -3119,11 +3124,100 @@ def generate(cls, config: SubsetConfig) -> Compose:
...
+    @classmethod
+    def _safe_eval(cls, str_to_eval: str) -> tuple[int, ...] | int:
+        """Safe eval function for _eval_input_size.
+
+        The function is implemented for `_eval_input_size`, so implementation is aligned to it as below
+        - Only multiplication or division evaluation are supported.
+        - Only constant and tuple can be operand.
+        - tuple is changed to numpy array before evaluation.
+        - result value is rounded to int.
+        """
```
Comment (goodsong81) [inline src/otx/core/data/transform_libs/torchvision.py]: To me, it looks like doing actually `_eval_input_size_str()`.

File: src/otx/core/data/transform_libs/torchvision.py [hunk 7]
```diff
@@ -3119,11 +3124,100 @@ def generate(cls, config: SubsetConfig) -> Compose:
...
+        Input size should be specified as $(input_size). (e.g. $(input_size) * 0.5)
+        Only simple multiplication or division evaluation is supported. For example,
+        $(input_size) * -0.5    => supported
+        $(input_size) * 2.1 / 3 => supported
+        $(input_size) + 1       => not supported
+        The function decides to pass tuple type or int type based on the type hint of the argument.
+        float point values are rounded to int.
+        """
+        if input_size is None:
+            return
```
Comment (sungchul1) [inline src/otx/core/data/transform_libs/torchvision.py]: It seems this condition is not passed because checking whether `input_size` is None is already in L3127.
I think we can use it and remove L3127.

Comment (goodsong81) [inline src/otx/core/data/transform_libs/torchvision.py]: And this function is doing `_(re)configure_input_size()` by filling the template if needed.

**Review by goodsong81 — DISMISSED** [review]
Quite nice feature for recipe users or developers.
Generally looks good, but it might be controversial if it's "too much".
Left minor comments from a code reader's viewpoint. Feel free to ignore.

File: src/otx/core/data/transform_libs/torchvision.py (hunk 6 above)
Comment (eunwoosh) [reply src/otx/core/data/transform_libs/torchvision.py]: yes, I think it's better to change function name. Thanks :)

File: src/otx/core/data/transform_libs/torchvision.py [hunk 8]
```diff
@@ -3119,11 +3124,100 @@ def generate(cls, config: SubsetConfig) -> Compose:
...
+        for key, val in cfg_transform["init_args"].items():
+            if not (isinstance(val, str) and "$(input_size)" in val):
+                continue
+            if model_cls is None:
+                model_cls = get_obj_from_str(cfg_transform["class_path"])
+
+            available_types = typing.get_type_hints(model_cls.__init__).get(key)
+            if available_types is None or check_type(_input_size, available_types):  # pass tuple[int, int]
+                cfg_transform["init_args"][key] = cls._safe_eval(
+                    val.replace("$(input_size)", f"({','.join(str(val) for val in _input_size)})"),
```
Comment (sungchul1) [inline src/otx/core/data/transform_libs/torchvision.py]: If the purpose is to get `(1, 1) * 2` from `val="$(input_size) * 2"` and `input_size=(1, 1)`, what about using `{input_size}` instead for below?
```suggestion
                    val.format(input_size=_input_size),
```

File: tests/unit/core/data/test_module.py (hunk 4 above)
Comment (eunwoosh) [reply tests/unit/core/data/test_module.py]: Nice comment. I'll add it for these cases

File: src/otx/core/utils/utils.py (hunk 1 above)
Comment (eunwoosh) [reply src/otx/core/utils/utils.py]: I see. I'll make it more readable

File: src/otx/core/data/transform_libs/torchvision.py [hunk 9]
```diff
@@ -3119,11 +3124,100 @@ def generate(cls, config: SubsetConfig) -> Compose:
...
+                )
+            elif check_type(_input_size[0], available_types):  # pass int
```
Comment (sungchul1) [inline src/otx/core/data/transform_libs/torchvision.py:3178]:
```suggestion
            elif all(check_type(_size, available_types) for _size in _input_size):  # pass int
```

File: src/otx/core/data/transform_libs/torchvision.py [hunk 10]
```diff
@@ -3119,11 +3124,100 @@ def generate(cls, config: SubsetConfig) -> Compose:
...
+        Only simple multiplication or division evaluation is supported. For example,
+        $(input_size) * -0.5    => supported
+        $(input_size) * 2.1 / 3 => supported
+        $(input_size) + 1       => not supported
+        The function decides to pass tuple type or int type based on the type hint of the argument.
+        float point values are rounded to int.
+        """
+        if input_size is None:
+            return
+        _input_size: tuple[int, int] = (input_size, input_size) if isinstance(input_size, int) else tuple(input_size)  # type: ignore[assignment]
```
Comment (sungchul1) [inline src/otx/core/data/transform_libs/torchvision.py]: Do we need to use this conversion from int to tuple?
I'm not sure if it is necessary because L3170 uses a single integer.

**Review by sungchul1 — COMMENTED** [review]
Thanks for your hard work. :)
I left some comments, please take a look.
And I think this PR has to include updated recipes to check unit and integration tests that are affected by this change.
What do you think?

File: src/otx/core/data/transform_libs/torchvision.py (hunk 2 above)
Comment (eunwoosh) [reply src/otx/core/data/transform_libs/torchvision.py:2372]: I just updated type hint for the input size feature, so I don't know well this. But when I checked the code, crop_size should be (H, W) because this class refers `ImageInfo` class and it has a tuple w/ (H, W) order. I'll update the unit test to use non-square value.

File: src/otx/core/data/transform_libs/torchvision.py (hunk 5 above)
Comment (eunwoosh) [reply src/otx/core/data/transform_libs/torchvision.py]: Thanks for pointing. I'll update it.

File: src/otx/core/data/transform_libs/torchvision.py (hunk 7 above)
Comment (eunwoosh) [reply src/otx/core/data/transform_libs/torchvision.py]: I think leaving the safeguard is better choice. How about leaving it as it is?

File: src/otx/core/data/transform_libs/torchvision.py (hunk 8 above)
Comment (eunwoosh) [reply src/otx/core/data/transform_libs/torchvision.py]: oh nice point. thanks

File: src/otx/core/data/transform_libs/torchvision.py (hunk 9 above)
Comment (eunwoosh) [reply src/otx/core/data/transform_libs/torchvision.py:3178]: should it check both `input_size[0]` and `input_size[1]`? I think it's ok to check just `input_size[0]` because `input_size[0]` is given as the argument. How do you think about it?

File: src/otx/core/data/transform_libs/torchvision.py (hunk 10 above)
Comment (eunwoosh) [reply src/otx/core/data/transform_libs/torchvision.py]: If we don't change it to tuple, then we should check what type `input_size` is currently because `input_size` can be either int or tuple. So, we need to add a few code to check current type of `input_size` after `if available_types is None or check_type(_input_size, available_types)`. For this reason, I think fixing input_size type to tuple makes code more readable.

File: src/otx/core/data/transform_libs/torchvision.py (hunk 7 above)
Comment (eunwoosh) [reply src/otx/core/data/transform_libs/torchvision.py]: done https://github.com/openvinotoolkit/training_extensions/pull/3759/commits/589c4c57264a27a5da2cfd9bb9aa3b2788c53bef

File: src/otx/core/data/transform_libs/torchvision.py (hunk 6 above)
Comment (eunwoosh) [reply src/otx/core/data/transform_libs/torchvision.py]: done https://github.com/openvinotoolkit/training_extensions/pull/3759/commits/589c4c57264a27a5da2cfd9bb9aa3b2788c53bef

File: src/otx/core/utils/utils.py (hunk 1 above)
Comment (eunwoosh) [reply src/otx/core/utils/utils.py]: done https://github.com/openvinotoolkit/training_extensions/pull/3759/commits/589c4c57264a27a5da2cfd9bb9aa3b2788c53bef

Comment (eunwoosh) [reply src/otx/core/utils/utils.py]: done https://github.com/openvinotoolkit/training_extensions/pull/3759/commits/589c4c57264a27a5da2cfd9bb9aa3b2788c53bef

File: src/otx/core/data/transform_libs/torchvision.py (hunk 2 above)
Comment (eunwoosh) [reply src/otx/core/data/transform_libs/torchvision.py:2372]: done https://github.com/openvinotoolkit/training_extensions/pull/3759/commits/2385183a05f5e4bc871e2c461287abcbf344008c

File: CHANGELOG.md (hunk 3 above)
Comment (eunwoosh) [reply CHANGELOG.md]: done https://github.com/openvinotoolkit/training_extensions/pull/3759/commits/080932be38ac791a29b6f3de922fa094bf58a7f3

File: src/otx/core/data/transform_libs/torchvision.py (hunk 5 above)
Comment (eunwoosh) [reply src/otx/core/data/transform_libs/torchvision.py]: done https://github.com/openvinotoolkit/training_extensions/pull/3759/commits/080932be38ac791a29b6f3de922fa094bf58a7f3

File: src/otx/core/data/transform_libs/torchvision.py (hunk 8 above)
Comment (eunwoosh) [reply src/otx/core/data/transform_libs/torchvision.py]: done https://github.com/openvinotoolkit/training_extensions/pull/3759/commits/080932be38ac791a29b6f3de922fa094bf58a7f3

File: src/otx/core/data/transform_libs/torchvision.py (hunk 7 above)
Comment (sungchul1) [reply src/otx/core/data/transform_libs/torchvision.py]: If so, let's keep this.

File: src/otx/core/data/transform_libs/torchvision.py (hunk 10 above)
Comment (sungchul1) [reply src/otx/core/data/transform_libs/torchvision.py]: I don't think about it.
Please check the below example.
```python
model_cls = get_obj_from_str("otx.core.data.transform_libs.torchvision.Resize") # <class 'otx.core.data.transform_libs.torchvision.Resize'>
key = "scale"
val = "$(input_size)"
available_types = typing.get_type_hints(model_cls.__init__).get(key) # int | tuple[int, int] | None

# for tuple
_input_size = (1, 1)
check_type(_input_size, available_types) # True with (1, 1)

# for int
_input_size = 1
check_type(_input_size, available_types) # True with 1

# for str
_input_size = "1"
check_type(_input_size, available_types) # False with TypeCheckError set at L3153

# both tuple and int can use the same argument for `cls._safe_eval`
cls._safe_eval(val.replace("$(input_size)", str(_input_size))
```
I think we don't need an additional condition you said because `check_type` can support both types.
If I'm wrong, please correct me :)

Comment (sungchul1) [reply src/otx/core/data/transform_libs/torchvision.py]: I thought it again, you seem to consider the cases about only single tuple argument or int argument, is it correct?
If so, I think your choice is more safe, please ignore it and let's keep this.

File: tests/unit/core/data/test_module.py (hunk 4 above)
Comment (eunwoosh) [reply tests/unit/core/data/test_module.py]: done https://github.com/openvinotoolkit/training_extensions/pull/3759/commits/8b3dc51d8e3b64d076e47c6bae6ea4d366b8a022
I added more unit test and change code to make proper error message if input_size is set to None and transform has `$(input_size)`.

File: src/otx/core/data/transform_libs/torchvision.py (hunk 10 above)
Comment (eunwoosh) [reply src/otx/core/data/transform_libs/torchvision.py]: yes right. I considered that case. Anyway thanks for thinking how to refine my code :)

**Review by harimkang — APPROVED** [review]

**Review by kprokofi — APPROVED** [review]
